umi_tools/count_tab.py

```diff
@@ -0,0 +1,131 @@
+'''
+count_tab.py - Count reads per gene from flatfile using UMIs
+=================================================================
+
+:Author: Ian Sudbery, Tom Smith
+:Release: $Id$
+:Date: |today|
+:Tags: Python UMI
+
+Purpose
+-------
+
+The purpose of this command is to count the number of reads per gene
+based on the read's gene assignment and UMI. Note this command is not
+currently able to perform per-cell counting. See the count command if
+you want to perform per-cell counting.
+
+The input must be in the following format (tab separated), where the
+first column is the read identifier and the second column is the
+assigned gene. The input must be sorted by the gene identifier:
+
+NS500668:144:H5FCJBGXY:2:22309:18356:15843_TCTAA    ENSG00000279457.3
+NS500668:144:H5FCJBGXY:3:23405:3971:19716_CGATG     ENSG00000225972.1
+
+You can perform any required file transformation and pipe the output
+directly to count_tab. For example to pipe output from featureCounts
+with the '-R' option you can do the following:
+
+    awk '$2=="Assigned" {print $1"\t"$3}' my.bam.featureCounts| sort -k2 |
+    umi_tools count_tab -S gene_counts.tsv -L count.log
+
+The tab file is assumed to contain each read id once only. For paired
+end reads with featureCounts you must include the "-p" option so each
+read id is included once only.
+
+'''
+
+import sys
+import collections
+import re
+
+# required to make iteritems python2 and python3 compatible
+from builtins import dict
+
+from functools import partial
+
+import pysam
+
+import pandas as pd
+import numpy as np
+
+try:
+    import umi_tools.Utilities as U
+except ImportError:
+    import Utilities as U
+
+try:
+    import umi_tools.network as network
+except ImportError:
+    import network
+
+try:
+    import umi_tools.umi_methods as umi_methods
+except ImportError:
+    import umi_methods
+
+# add the generic docstring text
+__doc__ = __doc__ + U.GENERIC_DOCSTRING
+
+
+def main(argv=None):
+    """script main.
+
+    parses command line options in sys.argv, unless *argv* is given.
+    """
+
+    if argv is None:
+        argv = sys.argv
+
+    # setup command line parser
+    parser = U.OptionParser(version="%prog version: $Id$",
+                            usage=globals()["__doc__"])
+
+    # add common options (-h/--help, ...) and parse command line
+    (options, args) = U.Start(parser, argv=argv, add_group_dedup_options=False)
+
+    if options.random_seed:
+        np.random.seed(options.random_seed)
+
+    if options.stdin != sys.stdin:
```

Comment:
If `options.stdin != sys.stdin`, then `options.stdin` should already be an open file handle pointing at the file specified by `-S` on the command line. 

image user runs `umi_tools count_tab -S feature_counts.tsv.gz`
Thus, with your code here on line 89 options.stdin is `<FileHandle: gzip <feature_counts.tsv.gz>>`. On line 90 you discover that it is not `stdin`, and so on line 91 you open the file `feature_counts.tsv.gz` again and set it to `infile`. 
Thus on line 95, `infile` contains what `options.stdin` contained on line 89, irrespective of what happens in 90-94, and the whole block could be replaced with a single line `infile=options.stdin`, or even `options.stdin` passed to `get_gene_count_tab` on line 104.
Bascially what